---
title: "Protein aggregation with protein age inferred from regression analysis "
author: "Ben Wilson"
output: html_document
---
  
Load mySQL table info for genes
```{r}
gene.data <- read.table("2017-06-20_EnsemblV73_MouseGenes.tab", sep="\t",header=T) 
protein.data <- read.table("2017-06-20_EnsemblV73_MouseProteins.tab", sep="\t",header=T)

APR.data <- merge(gene.data, protein.data, by.x="MouseGenesPrimaryKey", by.y="MouseGenesForeignKey")
remove(gene.data)
remove(protein.data)
```

Factorize gene family number and phylostratum
```{r}
APR.data$GeneFamilyNumber <- as.factor(APR.data$GeneFamilyNumber) #factorize gene family ID to prevent numeric interpretation by R
APR.data$GeneFamilyPhylostratum <- as.factor(APR.data$GeneFamilyPhylostratum) #factorize variable to ensure that it is not interpreted as numeric variable
APR.data$GeneFamilyPhylostratum <- with(APR.data, factor(GeneFamilyPhylostratum, levels = c(sort(unique(APR.data$GeneFamilyPhylostratum)),"Intergenic","Masked"))) # set levels of categorical phylostratum variable
APR.data$ExcludedGeneBinary <- as.factor(APR.data$ExcludedGeneBinary) #factorize binary category for gene families with ambiguous age

```

Set phylostrata for intergenic controls
```{r}
APR.data[which(APR.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl"),]$GeneFamilyPhylostratum <- "Intergenic" #Give intergenic sequence a categorical phylostratum for comparison
APR.data[which(APR.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),]$GeneFamilyPhylostratum <- "Masked" #Give repeat-masked intergenic sequence a categorical phylostratum

```

Map expression values from genes to scrambled controls
```{r}
match.vector <- match(APR.data[which(APR.data$GeneOrControlDesignation == "ScrambledByAminoAcidControl"),]$EnsemblGeneUID, APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),]$EnsemblGeneUID)
APR.data[which(APR.data$GeneOrControlDesignation == "ScrambledByAminoAcidControl"),]$PaxDBExpression <- APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),][match.vector,c("PaxDBExpression")]
```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, genes without rat homologs, and ORFan proteins
```{r}
APR.data <- APR.data[which(APR.data$GeneFamilyPhylostratum != "0" & APR.data$GeneFamilyPhylostratum != "20" & APR.data$ExcludedGeneBinary != "1" & !is.na(APR.data$EvolutionaryRateHighestOrthologRate)),] #exclude unclassifiable genes, false-positive rich ORFans
```

Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)

bcFit.TangoNumAA <- boxcoxfit(APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),]$TangoPentamerAAInAPRsDensity, lambda2 =  NULL) #Fit Box-Cox transform to each aggregation metric
lambda1.TangoNumAA <- unname(bcFit.TangoNumAA$lambda[1])
lambda1.TangoNumAA



bc.transform <- function(x,l1,l2){ #Create function to transform data according to fit parameters
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}


APR.data$bcTangoNumAA <- bc.transform(APR.data$TangoPentamerAAInAPRsDensity, lambda1.TangoNumAA, 0) 
```

Generate family of linear models for each phylostratum, test significance of gene vs. randomized control controlling for gene family variance
```{r}

#**********NLME FITS*********
library(nlme)

fit.APR.genes.family <- lme(bcTangoNumAA ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),], control = list(opt = "optim"), na.action = na.exclude, method = "ML")
#summary(fit.APR.genes.family) #fit linear mixed-effects model for genes with gene family as random error term

fit.APR.genes.family.expressed <- lme(bcTangoNumAA ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene" & !is.na(APR.data$PaxDBExpression)),], control = list(opt = "optim"), na.action = na.exclude, method = "ML")

fit.APR.genes.family.expression <- lme(bcTangoNumAA ~ GeneFamilyPhylostratum + PaxDBExpression, random = ~ 1|GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),], control = list(opt = "optim"), na.action = na.exclude, method = "ML")
anova(fit.APR.genes.family.expressed, fit.APR.genes.family.expression)
#summary(fit.APR.genes.family.expression) #fit linear mixed-effects model for genes with gene family as random error term

fit.APR.genes.family.length <- lme(bcTangoNumAA ~ GeneFamilyPhylostratum + ProteinSeqLength, random = ~ 1|GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),], control = list(opt = "optim"), na.action = na.exclude, method = "ML")
#summary(fit.APR.genes.family.length) #fit linear mixed-effects model for genes with gene family as random error term
anova(fit.APR.genes.family, fit.APR.genes.family.length)

fit.APR.genes.family.rate <- lme(bcTangoNumAA ~ GeneFamilyPhylostratum + EvolutionaryRateHighestOrthologRate, random = ~ 1|GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),], control = list(opt = "optim"), na.action = na.exclude, method = "ML")
anova(fit.APR.genes.family, fit.APR.genes.family.rate)


fit.APR.scrambled.family <- lme(bcTangoNumAA ~ GeneFamilyPhylostratum , random = ~ 1|GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "ScrambledByAminoAcidControl"),], control = list(opt = "optim"), na.action = na.exclude) #fit scrambled controls with linear mixed-effects model with gene family as random error term
fit.APR.intergenic <- lm(bcTangoNumAA ~ 1 , data = APR.data[which(APR.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl"),]) #fit standard linear model to intergenic controls
fit.APR.masked <- lm(bcTangoNumAA ~ 1 , data = APR.data[which(APR.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),]) #...same for repeat-masked controls

#remove(APR.data)
```

Import protein length and rate data from length/dNdS analysis
```{r}
protein.lengths <- readRDS("protein_lengths.rds")
length.lambda1 <- readRDS("length_lambda1.rds")
length.lambda2 <- readRDS("length_lambda2.rds")

evolutionary.rates <- readRDS("evolutionary_rates.rds")
rate.lambda1 <- readRDS("rates_lambda1.rds")
rate.lambda2 <- readRDS("rates_lambda2.rds")

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

protein.lengths$Lengths <- bc.reverse.transform(protein.lengths$Intercept,length.lambda1,length.lambda2)
vertebrate.lengths <- mean(protein.lengths$Lengths[2:length(protein.lengths$Lengths)])

evolutionary.rates$Rates <- bc.reverse.transform(evolutionary.rates$Intercept, rate.lambda1, rate.lambda2)
```

Plot means from each model with confidence intervals to test if they are significantly different
```{r}
# Put model estimates into data frames
last.index.genes <- length(fixed.effects(fit.APR.genes.family)) - 0 #store number of PS that have been fit (minus 2 for len and exp fixed effects)
genes.family.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.APR.genes.family)[1:last.index.genes]), Intercept = unname(fixed.effects(fit.APR.genes.family)[1:last.index.genes]), SE = unname(summary(fit.APR.genes.family)$tTable[,2][1:last.index.genes]), Model = "Genes") #create data frame to store lm results for genes
#remove(fit.APR.genes.family)

last.index.genes.rate <- length(fixed.effects(fit.APR.genes.family.rate)) - 1 #store number of PS that have been fit (minus 2 for len and exp fixed effects)
genes.family.rate.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.APR.genes.family.rate)[1:last.index.genes]), Intercept = unname(fixed.effects(fit.APR.genes.family.rate)[1:last.index.genes.rate]), SE = unname(summary(fit.APR.genes.family.rate)$tTable[,2][1:last.index.genes.rate]), Model = "Genes rate controlled") #create data frame to store lm results for genes
#remove(fit.APR.genes.family.rate)


last.index.scrambled <- length(fixed.effects(fit.APR.scrambled.family)) - 0 #store number of PS that have been fit for scrambled
scrambled.family.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.APR.scrambled.family)[1:last.index.scrambled]), Intercept = unname(fixed.effects(fit.APR.scrambled.family)[1:last.index.scrambled]), SE = unname(summary(fit.APR.scrambled.family)$tTable[,2][1:last.index.scrambled]), Model = "Scrambled") #create data frame to store lm results for the scrambled fit
remove(fit.APR.scrambled.family)

intergenic.data.frame <- data.frame(Phylostratum = c("Intergenic"), Intercept = summary(fit.APR.intergenic)$coefficients[1], SE = summary(fit.APR.intergenic)$coefficients[2], Model = "Intergenic") #create data frame to store intergenic fit

masked.data.frame <- data.frame(Phylostratum = c("Repeat-masked"), Intercept = summary(fit.APR.masked)$coefficients[1], SE = summary(fit.APR.masked)$coefficients[2], Model = "Repeat-masked") #create data frame to store repeat-masked intergenic fit


```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), combine data frames per categorization
```{r}
# Adjust intecepts, combine data frames
genes.family.data.frame$Intercept[2:last.index.genes] <- genes.family.data.frame$Intercept[2:last.index.genes] + genes.family.data.frame$Intercept[1] #increment means by first means to correct for R convention for storing fit results

scrambled.family.data.frame$Intercept[2:last.index.scrambled] <- scrambled.family.data.frame$Intercept[2:last.index.scrambled] + scrambled.family.data.frame$Intercept[1] #...same for scrambled

combined.data.frame.family <- data.frame(rbind(genes.family.data.frame, scrambled.family.data.frame, intergenic.data.frame, masked.data.frame)) #combine result data frames
combined.data.frame.family$Phylostratum <- replace(as.character(combined.data.frame.family$Phylostratum),1:(last.index.genes+last.index.scrambled),c(1:last.index.genes,1:last.index.scrambled)) #change PS names for the scrambled and gene fits to numeric characters
combined.data.frame.family$Phylostratum <- with(combined.data.frame.family, factor(Phylostratum, levels = c(1:max(last.index.genes,last.index.scrambled),"Intergenic","Repeat-masked"))) #change factor levels for PS for combined results data frame

```


Make plots to visualize effects of different categorization
```{r}
# Plots for each categorization
library("grid")
library("ggplot2")

#Define reverse transform
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

lambda1 <- lambda1.TangoNumAA
lambda2 <- 0
#Create colorblind friendly palette
colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

interval.times <- read.table("split_interval_times.txt")
axis.breaks <- -log10(interval.times$V1)
axis.breaks <- c(axis.breaks,-7.8,-7.75) #add extra points for controls
split.times <- as.character(round(interval.times$V1/1000000.0))

aggregation.summary.plot <- ggplot(combined.data.frame.family[which(combined.data.frame.family$Model != "Scrambled" ),], aes(colour = Model, fill = Model, shape = Model)) #Create ggplot object
aggregation.summary.plot <- aggregation.summary.plot + 
  geom_pointrange(aes(x = axis.breaks, y = bc.reverse.transform(Intercept, lambda1, lambda2), ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), size = 0.5, lwd = 1,   show.legend=FALSE) + #add point between error bars
  scale_x_continuous(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","*See figure\n legend","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","\nEuarchontoglires","Rodentia","Intergenic","Repeat-masked"), breaks = axis.breaks) + #change x-axis names of phylostrata to taxonomic split names
  geom_vline(xintercept = -7.85, color = "darkgrey", linetype = 2) +
  annotate("text", x = axis.breaks[c(1:5,11:19)], y = rep(-0.075,14), label = split.times[c(1:5,11:19)], angle=60, hjust = 0, vjust = 0.5, size =2.25) +
  annotate("text", x = -7.75, y = -0.07, label = "Avg. TMRCAs \n(mya)", size = 2.25 ) +
  scale_shape_manual(name = "", values = c(21,23,24)) + #manually set hollow shapes
  scale_colour_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue)) + #manually set line colors
  scale_fill_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue)) + #manually set fill colors
  ylab("Aggregation propensity") + #set y-axis label
  guides(shape=guide_legend(ncol=2, byrow=TRUE)) + #change shape of legend to two-column legend
  theme( 
    panel.grid.minor.x = element_blank(),
    axis.title.x = element_blank(), 
    axis.text.y = element_text(size=10),
    axis.title.y = element_text(size=11),
    axis.ticks = element_blank(),
    axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,13)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",11),colDarkBlue,colSkyBlue)),
    legend.title = element_text(size=11),
    legend.text = element_text(size=8),
    legend.background = element_rect(fill="transparent"),
    legend.position = c(0.21,0.83),
    legend.direction = "horizontal",
    plot.margin=unit(c(0.1,0.1,0.1,0.1), "cm")) #plot specifics for sizes, positions, orientations, etc.
aggregation.summary.plot

pdf(file = "APRvsPSGenesIntergenicMasked_TangoAANumInAPR.pdf", width =6.7, height = 3) #output plot to pdf
aggregation.summary.plot
dev.off()

```