---
title: "Regression analysis of hydrophobicity dispersion"
author: "Ben Wilson"
output: html_document
---
  
Load mySQL table info for genes
```{r}
gene.data <- read.table("2017-06-20_EnsemblV73_MouseGenes.tab", sep="\t",header=T) 
protein.data <- read.table("2017-06-20_EnsemblV73_MouseProteins.tab", sep="\t",header=T)

APR.data <- merge(gene.data, protein.data, by.x="MouseGenesPrimaryKey", by.y="MouseGenesForeignKey")
remove(gene.data)
remove(protein.data)
membrane.data <- read.table("EnsemblV73_MouseProteins_Membrane_Additions.tab", sep="\t",header=T)
APR.data <- merge(APR.data, membrane.data[,c("MouseProteinsPrimaryKey","MemProtKnown3DStructBinary")], by="MouseProteinsPrimaryKey")
```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, genes without rat homologs, and ORFan proteins
```{r}
APR.data <- APR.data[which(APR.data$GeneFamilyPhylostratum != "0" & APR.data$GeneFamilyPhylostratum != "20" & APR.data$ExcludedGeneBinary != "1" ),] #exclude unclassifiable genes, false-positive rich ORFans

```

Trim data frame to only the information needed
```{r}
genes <- APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),c("EnsemblGeneUID","GeneFamilyNumber","GeneFamilyPhylostratum","HydrophobicDispersionBlock6","EvolutionaryRateHighestOrthologRate","ProteinSeqLength","PaxDBExpression","MemProtKnown3DStructBinary")] #subset out genes
intergenic <- APR.data[which(APR.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl"),c("EnsemblGeneUID","GeneFamilyNumber","GeneFamilyPhylostratum","HydrophobicDispersionBlock6")] #subset out intergenic
masked <- APR.data[which(APR.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),c("EnsemblGeneUID","GeneFamilyNumber","GeneFamilyPhylostratum","HydrophobicDispersionBlock6")] #subset out intergenic
remove(APR.data) #clear large data frame with unnecessary information
```

Factorize categorical variables before linear model construction
```{r}
genes$GeneFamilyNumber <- as.factor(genes$GeneFamilyNumber) #factorize gene family identifier to prevent numeric interpretation
genes$GeneFamilyPhylostratum <- as.factor(genes$GeneFamilyPhylostratum) #factorize PS to prevent numeric interpretation in fit
genes$GeneFamilyPhylostratum <- with(genes, factor(GeneFamilyPhylostratum, levels = c(sort(unique(genes$GeneFamilyPhylostratum))))) # set levels of categorical phylostratum variable

```

Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)
bcFit <- boxcoxfit(genes$HydrophobicDispersionBlock6, lambda2 =  NULL) #Fit 1-parameter Box-Cox transformation
lambda1 <- unname(bcFit$lambda[1])
lambda2 <- 0
lambda1
lambda2

bc.transform <- function(x,l1,l2){
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}

genes$bcDispersion <- bc.transform(genes$HydrophobicDispersionBlock6, lambda1, lambda2) #transform data to reduce heteroskedasticity
intergenic$bcDispersion <- bc.transform(intergenic$HydrophobicDispersionBlock6, lambda1, lambda2) #transform data to reduce heteroskedasticity
masked$bcDispersion <- bc.transform(masked$HydrophobicDispersionBlock6, lambda1, lambda2) #transform data to reduce heteroskedasticity
```

Generate family of linear models for each phylostratum to give mean dispersion corrected for gene family non-independence
```{r}
library(nlme)

fit.dispersion.genes.standard <- lm(bcDispersion ~ GeneFamilyPhylostratum, data = genes)
AIC(fit.dispersion.genes.standard) #fit a standard linear model for comparison, print AIC

fit.dispersion.genes.family <- lme(bcDispersion ~ GeneFamilyPhylostratum , random = ~ 1|GeneFamilyNumber, data = genes, method = "ML")
summary(fit.dispersion.genes.family) #fit a mixed-effects model with gene family as a random effect term

fit.dispersion.genes.family.membrane <- lme(bcDispersion ~ GeneFamilyPhylostratum + MemProtKnown3DStructBinary, random = ~ 1|GeneFamilyNumber, data = genes, method = "ML")
summary(fit.dispersion.genes.family.membrane) #fit a mixed-effects model with gene family as a random effect term

fit.dispersion.genes.family.expressed <- lme(bcDispersion ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data = genes[which(!is.na(genes$PaxDBExpression)),], control = list(opt = "optim"), na.action = na.exclude, method = "ML")
summary(fit.dispersion.genes.family.expressed)

fit.dispersion.genes.family.expression <- lme(bcDispersion ~ GeneFamilyPhylostratum + PaxDBExpression, random = ~ 1|GeneFamilyNumber, data = genes, control = list(opt = "optim"), na.action = na.exclude, method = "ML")
summary(fit.dispersion.genes.family.expression)
anova(fit.dispersion.genes.family.expressed, fit.dispersion.genes.family.expression)
#summary(fit.APR.genes.family.expression) #fit linear mixed-effects model for genes with gene family as random error term

fit.dispersion.genes.family.length <- lme(bcDispersion ~ GeneFamilyPhylostratum + ProteinSeqLength, random = ~ 1|GeneFamilyNumber, data = genes, control = list(opt = "optim"), na.action = na.exclude, method = "ML")
summary(fit.dispersion.genes.family.length)
#summary(fit.APR.genes.family.length) #fit linear mixed-effects model for genes with gene family as random error term
anova(fit.dispersion.genes.family, fit.dispersion.genes.family.length)

fit.dispersion.genes.family.rate <- lme(bcDispersion ~ GeneFamilyPhylostratum + EvolutionaryRateHighestOrthologRate, random = ~ 1|GeneFamilyNumber, data = genes, control = list(opt = "optim"), na.action = na.exclude, method = "ML")
summary(fit.dispersion.genes.family.rate)
anova(fit.dispersion.genes.family, fit.dispersion.genes.family.rate)

fit.dispersion.intergenic <- lm(bcDispersion ~ 1, data = intergenic)
fit.dispersion.masked <- lm(bcDispersion ~ 1, data = masked)
#remove(genes) #clear data frame with gene information

```

Import protein length and rate data from length/dNdS analysis
```{r}
protein.lengths <- readRDS("protein_lengths.rds")
length.lambda1 <- readRDS("length_lambda1.rds")
length.lambda2 <- readRDS("length_lambda2.rds")

evolutionary.rates <- readRDS("evolutionary_rates.rds")
rate.lambda1 <- readRDS("rates_lambda1.rds")
rate.lambda2 <- readRDS("rates_lambda2.rds")


expression.levels <- readRDS("expression_levels.rds")
expression.lambda1 <- readRDS("expression_lambda1.rds")
expression.lambda2 <- readRDS("expression_lambda2.rds")

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

protein.lengths$Lengths <- bc.reverse.transform(protein.lengths$Intercept,length.lambda1,length.lambda2)
evolutionary.rates$Rates <- bc.reverse.transform(evolutionary.rates$Intercept, rate.lambda1, rate.lambda2)
expression.levels$ExpressionLevels <- bc.reverse.transform(expression.levels$Intercept, expression.lambda1, expression.lambda2)
```

Plot means from each model with confidence intervals, build data frames first
```{r}
# Put model estimates into data frames
last.index.genes <- length(fixed.effects(fit.dispersion.genes.family)) #get number of phylostrata that have been fit
genes.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.dispersion.genes.family)[1:last.index.genes]), Intercept = unname(fixed.effects(fit.dispersion.genes.family)[1:last.index.genes]), SE = unname(summary(fit.dispersion.genes.family)$tTable[,2][1:last.index.genes]), Model = "Genes") #create data frame to hold linear model results

# Put model estimates into data frames
last.index.length.corrected <- length(fixed.effects(fit.dispersion.genes.family.length)) - 1 #get number of phylostrata that have been fit
length.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.dispersion.genes.family.length)[1:last.index.length.corrected]), Intercept = unname(fixed.effects(fit.dispersion.genes.family.length)[1:last.index.length.corrected])+ unname(fixef(fit.dispersion.genes.family.length)[last.index.length.corrected + 1]) * protein.lengths$Lengths, SE = unname(summary(fit.dispersion.genes.family.length)$tTable[,2][1:last.index.length.corrected]), Model = "Length-corrected") #create data frame to hold linear model results

# Put model estimates into data frames
last.index.rate.corrected <- length(fixed.effects(fit.dispersion.genes.family.rate)) - 1 #get number of phylostrata that have been fit
rate.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.dispersion.genes.family.rate)[1:last.index.rate.corrected]), Intercept = unname(fixed.effects(fit.dispersion.genes.family.rate)[1:last.index.rate.corrected])+ unname(fixef(fit.dispersion.genes.family.rate)[last.index.rate.corrected + 1]) * evolutionary.rates$Rates, SE = unname(summary(fit.dispersion.genes.family.rate)$tTable[,2][1:last.index.rate.corrected]), Model = "Rate-corrected") #create data frame to hold linear model results

# Put model estimates into data frames
last.index.expression.corrected <- length(fixed.effects(fit.dispersion.genes.family.expression)) - 1 #get number of phylostrata that have been fit
expression.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.dispersion.genes.family.expression)[1:last.index.expression.corrected]), Intercept = unname(fixed.effects(fit.dispersion.genes.family.expression)[1:last.index.expression.corrected])+ unname(fixef(fit.dispersion.genes.family.expression)[last.index.expression.corrected + 1]) * expression.levels$ExpressionLevels, SE = unname(summary(fit.dispersion.genes.family.expression)$tTable[,2][1:last.index.expression.corrected]), Model = "Expression-corrected") #create data frame to hold linear model results


intergenic.data.frame <- data.frame(Phylostratum = c("Intergenic"), Intercept = summary(fit.dispersion.intergenic)$coefficients[1], SE = summary(fit.dispersion.intergenic)$coefficients[2], Model = "Intergenic") #create data frame to hold linear model results

masked.data.frame <- data.frame(Phylostratum = c("Repeat-masked"), Intercept = summary(fit.dispersion.masked)$coefficients[1], SE = summary(fit.dispersion.masked)$coefficients[2], Model = "Repeat-masked") #create data frame to hold linear model results


remove(fit.dispersion.intergenic)
remove(fit.dispersion.masked)

```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), factorize phylostratum
```{r}
# Adjust intecepts, combine data frames
genes.data.frame$Intercept[2:last.index.genes] <- genes.data.frame$Intercept[2:last.index.genes] + genes.data.frame$Intercept[1] #increment means by first mean
length.data.frame$Intercept[2:last.index.length.corrected] <- length.data.frame$Intercept[2:last.index.length.corrected] + length.data.frame$Intercept[1] #increment means by first mean
rate.data.frame$Intercept[2:last.index.rate.corrected] <- rate.data.frame$Intercept[2:last.index.rate.corrected] + rate.data.frame$Intercept[1] #increment means by first mean
expression.data.frame$Intercept[2:last.index.expression.corrected] <- expression.data.frame$Intercept[2:last.index.expression.corrected] + expression.data.frame$Intercept[1] #increment means by first mean

combined.data.frame <- rbind(genes.data.frame, intergenic.data.frame, masked.data.frame)
combined.data.frame$Phylostratum <- replace(as.character(combined.data.frame$Phylostratum),1:last.index.genes,c(1:last.index.genes)) #replace lm text with numbered categories
combined.data.frame$Phylostratum <- with(combined.data.frame, factor(Phylostratum, levels = c(1:last.index.genes,"Intergenic","Repeat-masked"))) # set levels of factorized PS

combined.df <- data.frame(rbind(genes.data.frame, length.data.frame, rate.data.frame, expression.data.frame)) #combine data frames for plotting
combined.df$Phylostratum <- replace(as.character(combined.df$Phylostratum),1:(last.index.genes+last.index.length.corrected+last.index.rate.corrected+last.index.expression.corrected),c(1:last.index.genes,1:last.index.length.corrected,1:last.index.rate.corrected,1:last.index.expression.corrected)) #replace linear model text with readable phylostratum names
combined.df$Phylostratum <- with(combined.df, factor(Phylostratum, levels = c(1:max(last.index.genes,last.index.length.corrected,last.index.rate.corrected,last.index.expression.corrected)))) #factorize phylostratum before plotting

```

Make plots to visualize effects of different categorization
```{r}
# Plots for each categorization
library("ggplot2")

# Function to reverse Box-Cox transformation to interpretable format
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

interval.times <- read.table("split_interval_times.txt")
axis.breaks <- -log10(interval.times$V1)
axis.breaks <- c(axis.breaks,-7.8,-7.75) #add extra points for controls


plot.dispersion <- ggplot(combined.data.frame, aes(colour = Model, fill = Model, shape = Model)) #create ggplot object
plot.dispersion <- plot.dispersion + 
  geom_pointrange(aes(x = axis.breaks, y = bc.reverse.transform(Intercept, lambda1, lambda2), ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), size = 0.5, lwd = 1, position = position_dodge(width = 0.05),  show.legend=FALSE) + #add points
  geom_hline(yintercept = 1, linetype = "dashed" ) + #add horizontal line for Poisson expectation
  scale_x_continuous(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","\nEuarchontoglires","Rodentia","Intergenic","Repeat-masked"), breaks = axis.breaks) + #change x-axis names of phylostrata to taxonomic split names
  geom_vline(xintercept = -7.85, color = "darkgrey", linetype = 2) +
  scale_colour_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue)) + #manually scale line colors
  scale_fill_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue)) + #manually scale fill colors
  scale_shape_manual(name = "", values = c(21,23,24)) + #manually set hollow shapes
  ylab("Hydrophobic clustering") + #set y-axis label
  theme(
    panel.grid.minor.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,13)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",11),colDarkBlue,colSkyBlue)),
        legend.title = element_text(size=11),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.4,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.1,0.1,0.1,0.1), "cm")) #plot specifics for sizes, positions, orientations, etc.

plot.dispersion

pdf(file = "DispersionGenesIntergenicMasked.pdf", width =6.7, height = 3) #output plot object to pdf
plot.dispersion
dev.off()



plot.dispersion.corrected <- ggplot(combined.df, aes(colour = Model, fill = Model, shape = Model)) #create ggplot object
plot.dispersion.corrected <- plot.dispersion.corrected + 
  geom_pointrange(aes(x = rep(axis.breaks[1:19],4), y = bc.reverse.transform(Intercept, lambda1, lambda2), ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), size = 0.5, lwd = 1, position = position_dodge(width = 0.05),  show.legend=TRUE) + #add points
  geom_hline(yintercept = 1, linetype = "dashed" ) + #add horizontal line for Poisson expectation
  scale_x_continuous(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","\nEuarchontoglires","Rodentia"), breaks = axis.breaks[1:19]) + #change x-axis names of phylostrata to taxonomic split names
  scale_colour_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue,colGreen), guide=guide_legend(nrow=2,byrow=TRUE)) + #manually scale line colors
  scale_fill_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue,colGreen)) + #manually scale fill colors
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #manually set hollow shapes
  #guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
  ylab("Hydrophobic clustering") + #set y-axis label
  theme(
    panel.grid.minor.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,11)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",11))),
        legend.title = element_text(size=11),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.4,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.1,0.1,0.1,0.1), "cm")) #plot specifics for sizes, positions, orientations, etc.

plot.dispersion.corrected

pdf(file = "DispersionGenesLengthRateExpression.pdf", width =6.7, height = 3) #output plot object to pdf
plot.dispersion.corrected
dev.off()


```