---
title: "Linear models for length and dNdS versus phylostratum"
author: "Ben Wilson"
output: html_document
---

Load mySQL table info for genes
```{r}
gene.data <- read.table("InputFinal/EnsemblV73_MouseGenes.tab", sep="\t",header=T) 
protein.data <- read.table("InputFinal/EnsemblV73_MouseProteins.tab", sep="\t",header=T)

APR.data <- merge(gene.data, protein.data, by.x="MouseGenesPrimaryKey", by.y="MouseGenesForeignKey")
remove(gene.data)
remove(protein.data)
```

Factorize gene family number and phylostratum
```{r}

paralogs.list <- c("ENSMUSG00000073631","ENSMUSG00000079800","ENSMUSG00000094728","ENSMUSG00000095240","ENSMUSG00000095366","ENSMUSG00000095450","ENSMUSG00000095787","ENSMUSG00000096100","ENSMUSG00000096850")
APR.data <- APR.data[which(APR.data$PS21_GeneFamilyPhylostratum != "0" & APR.data$PS21_GeneFamilyPhylostratum != "21" & APR.data$PS21_ExcludedGeneBinary != "1" ),] #exclude unclassifiable genes, false-positive rich ORFans
APR.data <- APR.data[which(!(APR.data$EnsemblGeneUID %in% paralogs.list)),]#remove genes with known paralogy to other genes


APR.data[which(APR.data$EnsemlGeneUID=="ENSMUSG00000067543"),]$PS21_GeneFamilyNumber <- APR.data[which(APR.data$EnsemlGeneUID=="ENSMUSG00000052262"),]$PS21_GeneFamilyNumber #Fix gene family for two pahari genes


APR.data$PS21_GeneFamilyNumber <- as.factor(APR.data$PS21_GeneFamilyNumber) #factorize gene family ID to prevent numeric interpretation by R
APR.data$PS21_GeneFamilyPhylostratum <- as.factor(APR.data$PS21_GeneFamilyPhylostratum) #factorize variable to ensure that it is not interpreted as numeric variable
APR.data$PS21_GeneFamilyPhylostratum <- with(APR.data, factor(PS21_GeneFamilyPhylostratum, levels = c(sort(unique(APR.data$PS21_GeneFamilyPhylostratum)),"Intergenic","Masked"))) # set levels of categorical phylostratum variable
APR.data$PS21_ExcludedGeneBinary <- as.factor(APR.data$PS21_ExcludedGeneBinary) #factorize binary category for gene families with ambiguous age

APR.data$ProteinSeqLength <- as.numeric(APR.data$ProteinSeqLength)
```

Trim data frames to only the information needed
```{r}
APR.data <- APR.data[,c("EnsemblGeneUID","PS21_GeneFamilyPhylostratum","PS21_GeneFamilyNumber","PaxDBExpression","GeneOrControlDesignation","EvolutionaryRateHighestOrthologRate","PS21_ExcludedGeneBinary", "ProteinSeqLength")] #subset needed columns

```


Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)
bcFit.length <- boxcoxfit(APR.data[APR.data$GeneOrControlDesignation == "CodingGene",]$ProteinSeqLength, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1.length <- unname(bcFit.length$lambda[1]) #extract transformation variables
lambda2.length <- unname(bcFit.length$lambda[2]) #...
lambda1.length #print transformation variables
lambda2.length #...

bc.transform <- function(x,l1,l2){ #function to transform data according to Box-Cox transformation fit
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}

APR.data$bcLength <- bc.transform(APR.data$ProteinSeqLength, lambda1.length, lambda2.length) #create another column variable for transformed  values

bcFit.rate <- boxcoxfit(APR.data[APR.data$GeneOrControlDesignation == "CodingGene",]$EvolutionaryRateHighestOrthologRate, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1.rate <- unname(bcFit.rate$lambda[1]) #extract transformation variables
lambda2.rate <- unname(bcFit.rate$lambda[2]) #...
lambda1.rate #print transformation variables
lambda2.rate #...

APR.data$bcRate <- bc.transform(APR.data$EvolutionaryRateHighestOrthologRate, lambda1.rate, lambda2.rate) #create another column variable for transformed values

bcFit.expression <- boxcoxfit(APR.data[APR.data$GeneOrControlDesignation == "CodingGene",]$PaxDBExpression, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1.expression <- unname(bcFit.expression$lambda[1]) #extract transformation variables
lambda2.expression <- unname(bcFit.expression$lambda[2]) #...
lambda1.expression #print transformation variables
lambda2.expression #...

APR.data$bcExpression <- bc.transform(APR.data$PaxDBExpression, lambda1.expression, lambda2.expression) #create another column variable for transformed values
```


Generate family of linear models for each phylostratum, test significance of gene vs. randomized control controlling for gene family variance
```{r}
library(nlme)


fit.length.genes <- lme(bcLength ~ 1 , random = ~ 1|PS21_GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),]) #generate a linear mixed-effects model fit to mouse genes with gene family as random error
summary(fit.length.genes) #print fit AIC for mouse genes with gene family random error

fit.rate.genes <- lme(bcRate ~ 1, random = ~ 1|PS21_GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),], na.action = "na.exclude") #generate a linear mixed-effects model fit to mouse genes with gene family as random error, add dNdS as fixed term, exclude genes without a corresponding dNdS value for mouse-rat comparisons
summary(fit.rate.genes) #print fit AIC for evolutionary rate corrected linear model

fit.expression.genes <- lme(bcExpression ~ 1, random = ~ 1|PS21_GeneFamilyNumber, data = APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),], na.action = "na.exclude") #generate a linear mixed-effects model fit to mouse genes with gene family as random error, add dNdS as fixed term, exclude genes without a corresponding dNdS value for mouse-rat comparisons
summary(fit.expression.genes) #print fit AIC for evolutionary rate corrected linear model

remove(APR.data)
```

Data frames for plots
```{r}
last.index.length <- length(fixed.effects(fit.length.genes))  #get number of phylostrata that have been fit
length.df <- data.frame(Phylostratum = names(fixed.effects(fit.length.genes)[1:last.index.length]), Intercept = unname(fixed.effects(fit.length.genes)[1:last.index.length]), SE = unname(summary(fit.length.genes)$tTable[,2][1:last.index.length]), Model = "Length") #build data frame to hold linear model data containing PS, mean bcLength, and SEM

last.index.rate <- length(fixed.effects(fit.rate.genes)) #get number of phylostrata that have been fit
rate.df <- data.frame(Phylostratum = names(fixef(fit.rate.genes)[1:last.index.rate]), Intercept = unname(fixef(fit.rate.genes)[1:last.index.rate]), SE = unname(summary(fit.rate.genes)$tTable[,2][1:last.index.rate]), Model = "Evolutionary rate") #build data frame to hold linear model data containing PS, mean and SEM

last.index.expression <- length(fixed.effects(fit.expression.genes)) #get number of phylostrata that have been fit
expression.df <- data.frame(Phylostratum = names(fixef(fit.expression.genes)[1:last.index.expression]), Intercept = unname(fixef(fit.expression.genes)[1:last.index.expression]), SE = unname(summary(fit.expression.genes)$tTable[,2][1:last.index.expression]), Model = "Expression level") #build data frame to hold linear model data containing PS, mean and SEM
#remove(fit.length.genes, fit.rate.genes) #remove linear model objects for genes
```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), combine data frames per categorization
```{r}
# Adjust intecepts, combine data frames
#length.df$Intercept[2:last.index.length] <- length.df$Intercept[2:last.index.length] + length.df$Intercept[1] #increment means by primary intercept, corrects for R convention of relative means
length.df$Phylostratum <- replace(as.character(length.df$Phylostratum),1:(last.index.length),c(1:last.index.length)) #replace linear model text with readable phylostratum names
length.df$Phylostratum <- with(length.df, factor(Phylostratum, levels = c(1:last.index.length)))#spcify factor levels

#rate.df$Intercept[2:last.index.rate] <- rate.df$Intercept[2:last.index.rate] + rate.df$Intercept[1]  #...same as above
rate.df$Phylostratum <- replace(as.character(rate.df$Phylostratum),1:(last.index.rate),c(1:last.index.rate)) #replace linear model text with readable phylostratum names
rate.df$Phylostratum <- with(rate.df, factor(Phylostratum, levels = c(1:last.index.rate)))#spcify factor levels

#expression.df$Intercept[2:last.index.expression] <- expression.df$Intercept[2:last.index.expression] + expression.df$Intercept[1]  #...same as above
expression.df$Phylostratum <- replace(as.character(expression.df$Phylostratum),1:(last.index.expression),c(1:last.index.expression)) #replace linear model text with readable phylostratum names
expression.df$Phylostratum <- with(expression.df, factor(Phylostratum, levels = c(1:last.index.expression)))#spcify factor levels
```

Get estimates for central tendencies
```{r}
#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

length.mean <- bc.reverse.transform(length.df$Intercept,lambda1.length,lambda2.length)
length.mean
rate.mean <- bc.reverse.transform(rate.df$Intercept, lambda1.rate, lambda2.rate)
rate.mean
expression.mean <- bc.reverse.transform(expression.df$Intercept, lambda1.expression, lambda2.expression)
expression.mean

```

Create colorblind friendly color scheme and reverse transform function
```{r}
library("ggplot2")
library("grid")

#Create colorblind friendly palette
colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

interval.times <- read.table("split_interval_times.txt")
axis.breaks <- -log10(interval.times$V1)
```

Generate plots for length/evolutionary-rate vs Phylostratum
```{r}
saveRDS(length.df, "protein_lengths.rds")
saveRDS(lambda1.length, "length_lambda1.rds")
saveRDS(lambda2.length, "length_lambda2.rds")


plot.length <- ggplot(length.df, aes(colour = Model, fill = Model, shape = Model)) #make ggplot plot object for linear model data frame, exclude mouse-specific proteins
plot.length <- plot.length + 
  geom_pointrange(aes(x = axis.breaks, y = bc.reverse.transform(Intercept, lambda1.length, lambda2.length), ymin = bc.reverse.transform(Intercept - SE, lambda1.length, lambda2.length), ymax = bc.reverse.transform(Intercept + SE, lambda1.length, lambda2.length)), size = 0.5, lwd = 0.5,show.legend =FALSE) + #add points for mean 
  scale_x_continuous(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","*See figure\n legend","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","\nEuarchontoglires","Rodentia"), breaks = axis.breaks) + #change x-axis names of phylostrata to taxonomic split names
  scale_colour_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue), guide = guide_legend(ncol=2, byrow=TRUE)) + #set colors for each category
  scale_fill_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue)) + #set fills for each category
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #set shapes for each category
  ylab("Protein length (aa)") + #set y-axis label
  theme( #set font sizes, orientations, justifications, etc.
        panel.grid.minor.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour = "black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,11)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",11))),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.8,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.1,0.1,0.1,0.1), "in"))

plot.length

#Create pdf with particular dimensions for journal
pdf(file = "LengthVsPhylostratum.pdf", width =6.7, height = 3)
plot.length
dev.off()

saveRDS(rate.df, "evolutionary_rates.rds")
saveRDS(lambda1.rate, "rates_lambda1.rds")
saveRDS(lambda2.rate, "rates_lambda2.rds")


plot.rate <- ggplot(rate.df, aes(colour = Model, fill = Model, shape = Model)) #make ggplot plot object for linear model data frame, exclude mouse-specific proteins
plot.rate <- plot.rate + 
  geom_pointrange(aes(x = axis.breaks, y = bc.reverse.transform(Intercept, lambda1.rate, lambda2.rate), ymin = bc.reverse.transform(Intercept - SE, lambda1.rate, lambda2.rate), ymax = bc.reverse.transform(Intercept + SE, lambda1.rate, lambda2.rate)), size = 0.5, lwd = 0.5,  show.legend =FALSE) + #add points for mean 
  scale_x_continuous(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","*See figure\n legend","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","\nEuarchontoglires","Rodentia"), breaks = axis.breaks) + #change x-axis names of phylostrata to taxonomic split names
  scale_colour_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue), guide = guide_legend(ncol=2, byrow=TRUE)) + #set colors for each category
  scale_fill_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue)) + #set fills for each category
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #set shapes for each category
  ylab("Evolutionary rate (dN/dS)") + #set y-axis label
  theme( #set font sizes, orientations, justifications, etc.
        panel.grid.minor.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,11)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",11))),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.33,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.1,0.1,0.1,0.1), "in"))


plot.rate

#Create pdf with particular dimensions for journal
pdf(file = "EvolutionaryRateVsPhylostratum.pdf", width =6.7, height = 3)
plot.rate
dev.off()

saveRDS(expression.df, "expression_levels.rds")
saveRDS(lambda1.expression, "expression_lambda1.rds")
saveRDS(lambda2.expression, "expression_lambda2.rds")


plot.expression <- ggplot(expression.df, aes(colour = Model, fill = Model, shape = Model)) #make ggplot plot object for linear model data frame, exclude mouse-specific proteins
plot.expression <- plot.expression + 
  geom_pointrange(aes(x = axis.breaks, y = bc.reverse.transform(Intercept, lambda1.expression, lambda2.expression), ymin = bc.reverse.transform(Intercept - SE, lambda1.expression, lambda2.expression), ymax = bc.reverse.transform(Intercept + SE, lambda1.expression, lambda2.expression)), size = 0.5, lwd = 0.5,  show.legend =FALSE) + #add points for mean 
  scale_x_continuous(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","*See figure\n legend","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","\nEuarchontoglires","Rodentia"), breaks = axis.breaks) + #change x-axis names of phylostrata to taxonomic split names
  scale_colour_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue), guide = guide_legend(ncol=2, byrow=TRUE)) + #set colors for each category
  scale_fill_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue)) + #set fills for each category
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #set shapes for each category
  ylab("Expression Level") + #set y-axis label
  theme( #set font sizes, orientations, justifications, etc.
        panel.grid.minor.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,11)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",11))),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.33,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.1,0.1,0.1,0.1), "in"))


plot.expression

#Create pdf with particular dimensions for journal
pdf(file = "ExpressionLevelVsPhylostratum.pdf", width =6.7, height = 3)
plot.expression
dev.off()


#Create stacked plot to show both correlations (shared x-axis)
ggGrobLength <- ggplotGrob(plot.length )
ggGrobRate <- ggplotGrob(plot.rate + theme(axis.text.x = element_blank()))
grid.draw(rbind(ggGrobRate,ggGrobLength,size="last"))
pdf(file = "RatePlusLengthVsPhylostratumShared.pdf", width = 6.7, height = 6)
grid.draw(rbind(ggGrobRate,ggGrobLength,size="last"))
dev.off()
```