---
title: "Regression analysis of SLiMs"
author: "Ben Wilson"
output: html_document
---
  
Load mySQL table info for genes
```{r}
gene.data <- read.table("2018-04-16_EnsemblV73_MouseGenes.tab", sep="\t",header=T) 
protein.data <- read.table("2018-04-16_EnsemblV73_MouseProteins.tab", sep="\t",header=T)

APR.data <- merge(gene.data, protein.data, by.x="MouseGenesPrimaryKey", by.y="MouseGenesForeignKey")
remove(gene.data)
remove(protein.data)
```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, genes without rat homologs, and ORFan proteins
```{r}
paralogs.list <- c("ENSMUSG00000073631","ENSMUSG00000079800","ENSMUSG00000094728","ENSMUSG00000095240","ENSMUSG00000095366","ENSMUSG00000095450","ENSMUSG00000095787","ENSMUSG00000096100","ENSMUSG00000096850")
APR.data <- APR.data[which(APR.data$PS21_GeneFamilyPhylostratum != "0" & APR.data$PS21_GeneFamilyPhylostratum != "21" & APR.data$PS21_ExcludedGeneBinary != "1" ),] #exclude unclassifiable genes, false-positive rich ORFans
APR.data <- APR.data[which(!(APR.data$EnsemblGeneUID %in% paralogs.list)),]#remove genes with known paralogy to other genes
APR.data[which(APR.data$EnsemlGeneUID=="ENSMUSG00000067543"),]$PS21_GeneFamilyNumber <- APR.data[which(APR.data$EnsemlGeneUID=="ENSMUSG00000052262"),]$PS21_GeneFamilyNumber #Fix gene family for two pahari genes


```

Trim data frame to only the information needed
```{r}
genes <- APR.data[which(APR.data$GeneOrControlDesignation == "CodingGene"),c("EnsemblGeneUID","PS21_GeneFamilyNumber","PS21_GeneFamilyPhylostratum", "ELMSLiMDensity")] #subset out genes
intergenic <- APR.data[which(APR.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl"),c("EnsemblGeneUID","PS21_GeneFamilyNumber","PS21_GeneFamilyPhylostratum","ELMSLiMDensity")] #subset out intergenic
masked <- APR.data[which(APR.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),c("EnsemblGeneUID","PS21_GeneFamilyNumber","PS21_GeneFamilyPhylostratum","ELMSLiMDensity")] #subset out intergenic
remove(APR.data) #clear large data frame with unnecessary information
```

Factorize categorical variables before linear model construction
```{r}
genes$PS21_GeneFamilyNumber <- as.factor(genes$PS21_GeneFamilyNumber) #factorize gene family identifier to prevent numeric interpretation
genes$PS21_GeneFamilyPhylostratum <- as.factor(genes$PS21_GeneFamilyPhylostratum) #factorize PS to prevent numeric interpretation in fit
genes$PS21_GeneFamilyPhylostratum <- with(genes, factor(PS21_GeneFamilyPhylostratum, levels = c(sort(unique(genes$PS21_GeneFamilyPhylostratum))))) # set levels of categorical phylostratum variable

```

Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)
bcFit <- boxcoxfit(genes$ELMSLiMDensity, lambda2 =  TRUE) #Fit 1-parameter Box-Cox transformation
lambda1 <- unname(bcFit$lambda[1])
lambda2 <- unname(bcFit$lambda[2])
lambda1
lambda2 

bc.transform <- function(x,l1,l2){
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}

genes$bcSLiMDensity <- bc.transform(genes$ELMSLiMDensity, lambda1, lambda2) #transform data to reduce heteroskedasticity
intergenic$bcSLiMDensity <- bc.transform(intergenic$ELMSLiMDensity, lambda1, lambda2) #transform data to reduce heteroskedasticity
masked$bcSLiMDensity <- bc.transform(masked$ELMSLiMDensity, lambda1, lambda2) #transform data to reduce heteroskedasticity
```

Generate family of linear models for each phylostratum to give mean SLiM density corrected for gene family non-independence
```{r}
library(nlme)

fit.SLiMDensity.genes.standard <- lm(ELMSLiMDensity ~ PS21_GeneFamilyPhylostratum, data = genes)
AIC(fit.SLiMDensity.genes.standard) #fit a standard linear model for comparison, print AIC
fit.SLiMDensity.genes <- lme(ELMSLiMDensity ~ PS21_GeneFamilyPhylostratum , random = ~ 1|PS21_GeneFamilyNumber, data = genes)
summary(fit.SLiMDensity.genes) #fit a mixed-effects model with gene family as a random effect term

fit.SLiMDensity.intergenic <- lm(ELMSLiMDensity ~ 1, data = intergenic)
fit.SLiMDensity.masked <- lm(ELMSLiMDensity ~ 1, data = masked)
#remove(genes) #clear data frame with gene information

```

Plot means from each model with confidence intervals, build data frames first
```{r}
# Put model estimates into data frames
last.index.genes <- length(fixed.effects(fit.SLiMDensity.genes)) #get number of phylostrata that have been fit
genes.data.frame <- data.frame(Phylostratum = names(fixed.effects(fit.SLiMDensity.genes)[1:last.index.genes]), Intercept = unname(fixed.effects(fit.SLiMDensity.genes)[1:last.index.genes]), SE = unname(summary(fit.SLiMDensity.genes)$tTable[,2][1:last.index.genes]), Model = "Genes") #create data frame to hold linear model results

intergenic.data.frame <- data.frame(Phylostratum = c("Intergenic"), Intercept = summary(fit.SLiMDensity.intergenic)$coefficients[1], SE = summary(fit.SLiMDensity.intergenic)$coefficients[2], Model = "Intergenic") #create data frame to hold linear model results

masked.data.frame <- data.frame(Phylostratum = c("Repeat-masked"), Intercept = summary(fit.SLiMDensity.masked)$coefficients[1], SE = summary(fit.SLiMDensity.masked)$coefficients[2], Model = "Repeat-masked") #create data frame to hold linear model results

remove(fit.SLiMDensity.genes) #clear fit object
remove(fit.SLiMDensity.intergenic)
remove(fit.SLiMDensity.masked)

```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), factorize phylostratum
```{r}
# Adjust intecepts, combine data frames
genes.data.frame$Intercept[2:last.index.genes] <- genes.data.frame$Intercept[2:last.index.genes] + genes.data.frame$Intercept[1] #increment means by first mean
combined.data.frame <- rbind(genes.data.frame, intergenic.data.frame, masked.data.frame)
combined.data.frame$Phylostratum <- replace(as.character(combined.data.frame$Phylostratum),1:last.index.genes,c(1:last.index.genes)) #replace lm text with numbered categories
combined.data.frame$Phylostratum <- with(combined.data.frame, factor(Phylostratum, levels = c(1:last.index.genes,"Intergenic","Repeat-masked"))) # set levels of factorized PS

```

Make plots to visualize effects of different categorization
```{r}
# Plots for each categorization
library("ggplot2")

# Function to reverse Box-Cox transformation to interpretable format
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

interval.times <- read.table("split_interval_times.txt")
axis.breaks <- -log10(interval.times$V1)
#axis.breaks <- c(axis.breaks,-7.8,-7.75) #add extra points for controls

log.scale <- c(axis.breaks,-6.9,-6.85)
label.vector <- c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","","","","","","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria\n","Euarchontoglires","Rodentia","Mus","Intergenic","Masked")
split.times <- as.character(round(interval.times$V1/1000000.0))


plot.SLiMDensity <- ggplot(combined.data.frame, aes(colour = Model, fill = Model, shape = Model)) #create ggplot object
plot.SLiMDensity <- plot.SLiMDensity + 
  geom_pointrange(aes(x = log.scale, y = Intercept, ymin = Intercept - SE, ymax =Intercept + SE), size = 0.5, lwd = 1,   show.legend=FALSE) + #add points
  scale_x_continuous(labels=label.vector, breaks = log.scale) + #change x-axis names of phylostrata to taxonomic split names
  geom_vline(xintercept = -7, color = "darkgrey", linetype = 2) +
  scale_colour_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue)) + #manually scale line colors
  scale_fill_manual(name = "", values = c(colBlack,colDarkBlue,colSkyBlue)) + #manually scale fill colors
  scale_shape_manual(name = "", values = c(21,23,24)) + #manually set hollow shapes
  ylab("SLiM density") + #set y-axis label
  theme(
    panel.grid.minor.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=c(rep(9,7),8,rep(9,14)), hjust = 1, vjust = 0.5, colour=c(rep("black",7),"lightslategrey",rep("black",12),colDarkBlue,colSkyBlue)),
        legend.title = element_text(size=11),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.4,0.78),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.1,0.1,0.1,0.1), "cm")) #plot specifics for sizes, positions, orientations, etc.

plot.SLiMDensity


tiff(file = "SLiMs_Intergenic_Masked.tif", width =20, height = 10, units = "cm", res = 300) #output plot object to tif
plot.SLiMDensity
dev.off()


```